### DISCUSSION TITLE
Parallel simulations - SWMM calibration

### DISCUSSION THREAD

**Discussion by tneivade** [pyswmm/pyswmm#582] [Q&A] [answered]
Greetings! 
My name is Thiago, I am a graduate student, and I am working on a model calibration in a watershed of 19 sq miles. I do have observed data for depth in 2 stations and in one of them also for flow discharge. 

I started my sensitivity analysis for the parameters using pyswmm for better documentation of comparison plots, error metrics, reports and modeled timeseries for nodes and links I am comparing with observed data. 

However, when I started running the model calibration using PCSWMM SRTC tool, it took 40 minutes to run a simulation and 40 more to switch parameter and generate a scenario, When I ran using PYSWMM, it took about 4h30min per simulation. Despite the inp file being set to use 6 threads out of 8, I think by standard it only uses 1 processor, right? 

I tried changing a few parameters (time steps - wet/ dry / routing) but I could not change significantly because my errors for continuity of the model got way too big.

I read some topics about it here in this forum and I tried also a few suggestions, but none of them change substantially the duration for each run. I see some comments upon parallel simulations. I am still a beginner on python coding so I didn't figure it out what would be necessary to run simulations in parallel. Are there any libraries, examples or tutorials that could help me on that?

Also, I see there is a way of running EPA SWMM Engine/Open SWMM using linux. My second idea would be using a supercomputer to run my scripts of pyswmm. Any of you had done anything similar and could enlighten me on where to look for documentation and materials about it? If I got these multiple simulations running, it would be already a great help!

I appreciate in advance any contributions!

**Comment by bemcdonnell** [comment]
@tneivade, It would be helpful to see some of your code/models.  I'm curious if there is a better way to write your Python code to run faster.  PySWMM is actually faster than EPASWMM by a small margin.  This leads me to believe that it could be your interfacing code.  PySWMM also takes advantage of threading.  

RE: your question about linux, PySWMM is fully supported on linux and if you head down that direction, it wouldn't be a bad idea.  Again, I think it is worth investigating how you are integrating with pyswmm.

**Reply by tneivade** [reply]
Hello, tks for the quick reply! I will paste it here my general settings for the simulation and my script (configure and running simulation).

  ****************
  Analysis Options
  ****************
  Flow Units ............... CFS
  Process Models:
    Rainfall/Runoff ........ YES
    RDII ................... NO
    Snowmelt ............... NO
    Groundwater ............ NO
    Flow Routing ........... YES
    Ponding Allowed ........ YES
    Water Quality .......... NO
  Infiltration Method ...... GREEN_AMPT
  Flow Routing Method ...... DYNWAVE
  Surcharge Method ......... EXTRAN
  Starting Date ............ 08/07/2024 22:00:00
  Ending Date .............. 08/12/2024 06:00:00
  Antecedent Dry Days ...... 0.0
  Report Time Step ......... 00:05:00
  Wet Time Step ............ 00:05:00
  Dry Time Step ............ 00:15:00
  Routing Time Step ........ 0.60 sec
  Variable Time Step ....... YES
  Maximum Trials ........... 30
  Number of Threads ........ 6
  Head Tolerance ........... 0.010000 ft

#Preset where I import the libraries, calling the paths for csv observed data, the inp file and the output directory, define error metrics and simulation period._

_import os
import shutil
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from swmm_api.input_file import read_inp_file
from pyswmm import Simulation, Nodes, Links

base_inp = "SWMMrun.inp"
base_dir = ".../4MR_PCSWMM_model/"
base_model_path = os.path.join(base_dir, base_inp)

observed_csvs = {
    ".../12003.csv",
    ".../20907.csv,
    ".../21606.csv"
}

output_dir = r"./outputs_uncertainty_analysis/"
os.makedirs(output_dir, exist_ok=True)

# Simulation period
start_time = datetime(2024, 8, 7, 22)
end_time = datetime(2024, 8, 12, 6)

# Node/Link IDs
target_elements = {
    "12003": ("depth", "node"),
    "20907": ("depth", "node"),
    "21606": ("flow", "link")  # force it to use the conduit
}

def compute_metrics(sim, obs):
    mask = ~np.isnan(sim) & ~np.isnan(obs)
    sim, obs = sim[mask], obs[mask]
    mse = mean_squared_error(obs, sim)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(obs, sim)
    nse = 1 - (np.sum((sim - obs)**2) / np.sum((obs - np.mean(obs))**2))
    bias = np.mean(sim - obs)
    return {"RMSE": rmse, "NSE": nse, "MAE": mae, "BIAS": bias}_

_# === Run simulation and collect results ===_
results = {}

with Simulation(base_model_path) as sim:
    sim.start_time = start_time
    sim.end_time = end_time

    nodes = Nodes(sim)
    links = Links(sim)

    times, data = [], {k: [] for k in target_elements}
    for step in sim:
        times.append(sim.current_time)
        for elem_id, (attr, typ) in target_elements.items():
            if typ == "link":
                val = getattr(links[elem_id], attr)
            elif typ == "node":
                val = getattr(nodes[elem_id], attr)
            else:
                val = np.nan
            data[elem_id].append(val)

**Reply by bemcdonnell** [reply, marked as answer]
@tneivade, this isn't enough information for me to run benchmark tests on performance.  However, I do have a few thoughts that could help you:

- it looks like you're appending values to an array but you're also querying depth a just a couple locations.  Consider running your simulation without trying to get results out, then using the Output module to pull out the timeseries.  This will let your simulation run un-interrupted.  This is documented well inside `pyswmm`.  Doing this will let you remove the inefficient append list.
```python
with Simulation(....) as sim:
    for step in sim:
        pass

with Output() as out:
    # query from the output file the depth for those objects
``` 
- You're running your model with a 0.6 second routing time step.  Is this for stability?  I suggest 30 seconds and reducing if you need
- it you want to decrease the frequency which pyswmm "interacts" with swmm, use `sim.step_advance(300)`.  This will for sure speed things up.
```python
with Simulation(....) as sim:
    sim.step_advance(300) #seconds
    for step in sim:
        pass
```

**Reply by tneivade** [reply]
The original setup was for 0.5 seconds on routing time step. I started to increase this variable, but the error was getting bigger so I went back and stopped changing it.

I will try both suggestions (running only the simulation and get other things from output and using the step_advance, Later on I will bring feedback. Thank you!

**Reply by tneivade** [reply]
I still didn't run  pyswmm and swmm engine for linux, but the suggestion for just executing the simulation and extracting timeseries nodes/ links and error metrics after using the output was really helpful. That really speed up the simulations (around 50 mins now!). Thanks again!
